logagg_collector/collector.py

```diff
@@ -0,0 +1,926 @@
+    QUEUE_READ_TIMEOUT = 1 # Wait time when doing blocking read on the in-mem q
+    PYGTAIL_ACK_WAIT_TIME = 0.05 # TODO: Document this
+    SCAN_FPATTERNS_INTERVAL = 30 # How often to scan filesystem for files matching fpatterns
+    HEARTBEAT_RESTART_INTERVAL = 30 # Wait time if heartbeat sending stops
+    LOGAGGFS_FPATH_PATTERN = re.compile("[a-fA-F\d]{32}") # MD5 hash pattern
+    SERVERSTATS_FPATH = '/var/log/serverstats/serverstats.log' # Path to serverstats log file
+    DOCKER_FORMATTER = 'logagg_collector.formatters.docker_file_log_driver' # Formatter name for docker_file_log_driver
+
+    LOG_STRUCTURE = {
+        'id': basestring,
+        'timestamp': basestring,
+        'file' : basestring,
+        'host': basestring,
+        'formatter' : basestring,
+        'raw' : basestring,
+        'type' : basestring,
+        'level' : basestring,
+        'event' : basestring,
+        'data' : dict,
+        'error' : bool,
+        'error_tb' : basestring,
+    }
+
+
+    def __init__(self, host, port, master, data_dir, logaggfs_dir, log=utils.DUMMY):
+
+        self.host = host
+        self.port = port
+        self.master = master
+
+        # For storing state
+        data_path = os.path.abspath(os.path.join(data_dir, 'logagg-data'))
+        self.data_path = utils.ensure_dir(data_path)
+
+        self.master = master
+
+        self.log = log
+
+        # For remembering the state of files
+        self.state = DiskDict(self.data_path)
+
+        # Initialize logaggfs paths
+        self.logaggfs = self._init_logaggfs_paths(logaggfs_dir)
+
+        # Log fpath to thread mapping
+        self.log_reader_threads = {}
+        # Handle name to formatter fn obj map
+        self.formatters = {}
+        self.queue = queue.Queue(maxsize=self.QUEUE_MAX_SIZE)
+
+        # Add initial files i.e. serverstats to state
+        if not self.state['fpaths']:
+            self.log.info('init_fpaths')
+            self._init_fpaths()
+
+        # register self to master
+        self.register_to_master()
+
+        # Create nsq_sender
+        self.nsq_sender_logs, self.nsq_sender_heartbeat = self._init_nsq_sender()
+
+        self._ensure_trackfiles_sync()
+
+
+    def register_to_master(self):
+        '''
+        Request authentication with master details
+        
+        Sample url: http://localhost:1088/logagg/v1/register_component?namespace=master
+                    &topic_name=logagg&host=localhost&port=1088
+        '''
+        #TODO: test case
+        master = self.master
+        url = self.REGISTER_URL.format(master_address=master.host+':'+master.port,
+                namespace=self.NAMESPACE,
+                topic_name=master.topic_name,
+                host=self.host,
+                port=self.port)
+
+        try:
+            register = requests.get(url)
+            register_result = json.loads(register.content.decode('utf-8'))
+            if register_result['success']:
```

Comment:
I believe register_result is a dict which will cause the problem if the URL is not reachable. please do use register_result.get("success", False). Please do get habituated to this kind of approach when we are seeking data from another source.